Support running tests with pytest

Some people like to use pytest as their test runner. Leaving aside discussions about whether those people are misguided or not, I noticed that the Traits test suite currently fails catastrophically under pytest. If the fixes needed to make it work out of the box under `pytest` are small, we should consider making them.

Here's the current result of doing a `python -m pip install -e .` followed by `pytest` in a freshly-created Python 3.8 venv:

```
mdickinson@mirzakhani traits % python -m venv --clear ~/.venvs/traits && source ~/.venvs/traits/bin/activate
(traits) mdickinson@mirzakhani traits % pip install -q --upgrade pip                                                 
(traits) mdickinson@mirzakhani traits % pip install -q pytest                                                        
(traits) mdickinson@mirzakhani traits % pip install -q -e .                                                          
(traits) mdickinson@mirzakhani traits % pytest
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /Users/mdickinson/Enthought/ETS/traits
collected 1419 items / 1 error / 1418 selected                                                                                                               

=========================================================================== ERRORS ===========================================================================
________________________________________________ ERROR collecting traits-stubs/traits_stubs_tests/test_all.py ________________________________________________
ImportError while importing test module '/Users/mdickinson/Enthought/ETS/traits/traits-stubs/traits_stubs_tests/test_all.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
traits-stubs/traits_stubs_tests/test_all.py:17: in <module>
    from traits_stubs_tests.util import MypyAssertions
traits-stubs/traits_stubs_tests/util.py:18: in <module>
    from mypy import api as mypy_api
E   ModuleNotFoundError: No module named 'mypy'
====================================================================== warnings summary ======================================================================
traits/testing/tests/test_unittest_tools.py:41
  /Users/mdickinson/Enthought/ETS/traits/traits/testing/tests/test_unittest_tools.py:41: PytestCollectionWarning: cannot collect test class 'TestObject' because it has a __init__ constructor (from: traits/testing/tests/test_unittest_tools.py)
    class TestObject(HasTraits):

traits/tests/test_property_notifications.py:38
  /Users/mdickinson/Enthought/ETS/traits/traits/tests/test_property_notifications.py:38: PytestCollectionWarning: cannot collect test class 'Test' because it has a __init__ constructor (from: traits/tests/test_property_notifications.py)
    class Test(HasTraits):

traits/tests/test_property_notifications.py:54
  /Users/mdickinson/Enthought/ETS/traits/traits/tests/test_property_notifications.py:54: PytestCollectionWarning: cannot collect test class 'Test_1' because it has a __init__ constructor (from: traits/tests/test_property_notifications.py)
    class Test_1(Test):

traits/util/tests/test_record_events.py:28
  /Users/mdickinson/Enthought/ETS/traits/traits/util/tests/test_record_events.py:28: PytestCollectionWarning: cannot collect test class 'TestObject' because it has a __init__ constructor (from: traits/util/tests/test_record_events.py)
    class TestObject(HasTraits):

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================================== short test summary info ===================================================================
ERROR traits-stubs/traits_stubs_tests/test_all.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================ 4 warnings, 1 error in 1.07s ================================================================
```

The warnings look easy to fix by simply renaming. The `mypy` error could (and should) be fixed by making `mypy` an optional dependency in the same way that we do for Sphinx and other bits and pieces. With `mypy` installed, I get just one failure:

```
E               AssertionError: 
E               /Users/mdickinson/Enthought/ETS/traits/traits-stubs/traits_stubs_tests/examples/Time.py
E               /var/folders/07/jbbjv8b53bs5y9xyjdyn1zgw0000gn/T/tmpmr0r37g0/Time.py:13: error: Too few arguments for "datetime"  [call-arg]
E               Found 1 error in 1 file (checked 1 source file)
```

mypy dependency fixed in #1289.

For the answer to @kitchoi's question:

> what would be the outcome if we do `python -m pip install -e .[test]` instead?

On current master (so including the #1289 fix), after:

```
mdickinson@mirzakhani traits % python -m venv --clear ~/.venvs/traits && source ~/.venvs/traits/bin/activate
(traits) mdickinson@mirzakhani traits % pip install -q --upgrade pip setuptools wheel
(traits) mdickinson@mirzakhani traits % pip install -q pytest
(traits) mdickinson@mirzakhani traits % pip install -q -e ".[test]"
```

I get the same 4 warnings as before about non-test classes named "Test*", one mypy failure, various `DeprecationWarning` s, but otherwise a clean, complete test run:

```
============================================= short test summary info =============================================
FAILED traits-stubs/traits_stubs_tests/test_all.py::TestAnnotations::test_all - AssertionError: 
================================== 1 failed, 1419 passed, 39 warnings in 43.22s ===================================
```

The `mypy` test failure appears to be to do with the way that packages have been installed: it looks as though the traits stubs aren't being picked up:

```
E               AssertionError: 
E               /Users/mdickinson/Enthought/ETS/traits/traits-stubs/traits_stubs_tests/examples/Time.py
E               /var/folders/07/jbbjv8b53bs5y9xyjdyn1zgw0000gn/T/tmpax4g4yd6/Time.py:13: error: Too few arguments for "datetime"  [call-arg]
E               Found 1 error in 1 file (checked 1 source file)
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -10,3 +10,6 @@ target-version = ['py36']
 profile = 'black'
 line_length = 79
 order_by_type = 'False'
+
+[tool.pytest.ini_options]
+addopts = '--ignore=traits-stubs'

